graders/python/python_autograder/pl_execute.py

```diff
@@ -51,20 +60,24 @@ def execute_code(fname_ref, fname_student, include_plt=False,
         str_setup = f.read()
     with open(fname_ref, 'r', encoding='utf-8') as f:
         str_ref = f.read()
-    try:
-        with open(join(filenames_dir, 'leading_code.py'), 'r', encoding='utf-8') as f:
-            str_leading = f.read()
-    except:
-        str_leading = ''
+
+    # Read in leading, trailing code
+    str_leading = try_read(join(filenames_dir, 'leading_code.py'))
+    str_trailing = try_read(join(filenames_dir, 'trailing_code.py'))
+
+    # Read student code (and transform if necessary) and append leading/trailing code
     with open(fname_student, 'r', encoding='utf-8') as f:
         filename, extension = splitext(fname_student)
         if extension == '.ipynb':
-            str_student = str_leading + pl_helpers.extract_ipynb_contents(f, ipynb_key)
+            str_student = pl_helpers.extract_ipynb_contents(f, ipynb_key)
         else:
             str_student = f.read()
+    str_student = str_leading + str_student + str_trailing
+
     with open(join(filenames_dir, 'test.py'), encoding='utf-8') as f:
         str_test = f.read()
 
+    # Delete sensitive code so students can't read e.g. test cases or setup code
```

Comment:
Should we also delete leading/trailing code? I know that we will often want to show then to students, by maybe we want to hide them sometimes? In any case, I don't think that deleting them can do any harm?

Reply:
I agree that deleting makes sense. The code can always be shown like in the example, using pl-code